### ISSUE TITLE
LSF affinity files override command line arguments

### ISSUE THREAD

**Issue by zerothi** [open-mpi/ompi#6631]
### What version of Open MPI are you using? (e.g., v1.10.3, v2.1.0, git branch name and hash, etc.)
3.1.3

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

- GCC 8.2.0
- Configure command line (from distribution tarball) (from `ompi_info`):
```bash
'CC=gcc' 'CXX=g++' 'FC=gfortran' '--prefix=/appl/gcc/8.2.0/openmpi/3.1.3-lsf10b' '--with-lsf' '--without-tm' '--enable-mpi-cxx' '--enable-cxx-exceptions' '--with-hwloc=/appl/gcc/8.2.0/openmpi/3.1.3-lsf10b' '--with-threads=posix' '--enable-orterun-prefix-by-default'
```

### Please describe the system on which you are running

* Operating system/version: 
ScientificLinux 7.3 (should be unrelated)
* Computer hardware: 
XeonE5-2650v4, 2 sockets, 12 cores per socket.
* Network type: 
Unrelated

-----------------------------

## Details of the problem

The problem is that starting a program with explicit command line arguments detailing mapping and binding is not enforced when the `LSB_AFFINITY_HOSTFILE` is set.
This is a bit surprising to new users and I would suggest that command line has precedence.

I'll give a few examples and the resulting binding:

- LSF script:
```bash
#BSUB -n 24
export OMP_NUM_THREADS=6
mpirun -n 4 --map-by ppr:2:socket,PE=$OMP_NUM_THREADS -x OMP_NUM_THREADS --report-bindings ./run.sh
```
And I get this binding:
```bash
[n-62-21-38:19485] MCW rank 0 bound to socket 0[core 0[hwt 0]]: [B/././././././././././.][./././././././././././.]
[n-62-21-38:19485] MCW rank 1 bound to socket 0[core 1[hwt 0]]: [./B/./././././././././.][./././././././././././.]
[n-62-21-38:19485] MCW rank 2 bound to socket 0[core 2[hwt 0]]: [././B/././././././././.][./././././././././././.]
[n-62-21-38:19485] MCW rank 3 bound to socket 0[core 3[hwt 0]]: [./././B/./././././././.][./././././././././././.]
```
which is wrong. However, with
```bash
#BSUB -n 24
unset LSB_AFFINITY_HOSTFILE
export OMP_NUM_THREADS=6
mpirun -n 4 --map-by ppr:2:socket,PE=$OMP_NUM_THREADS -x OMP_NUM_THREADS --report-bindings ./run.sh
```
I get:
```bash
[n-62-21-38:19497] MCW rank 0 bound to socket 0[core 0[hwt 0]], socket 0[core 1[hwt 0]], socket 0[core 2[hwt 0]], socket 0[core 3[hwt 0]], socket 0[core 4[hwt 0]], socket 0[core 5[hwt 0]]: [B/B/B/B/B/B/./././././.][./././././././././././.]
[n-62-21-38:19497] MCW rank 1 bound to socket 0[core 6[hwt 0]], socket 0[core 7[hwt 0]], socket 0[core 8[hwt 0]], socket 0[core 9[hwt 0]], socket 0[core 10[hwt 0]], socket 0[core 11[hwt 0]]: [././././././B/B/B/B/B/B][./././././././././././.]
[n-62-21-38:19497] MCW rank 2 bound to socket 1[core 12[hwt 0]], socket 1[core 13[hwt 0]], socket 1[core 14[hwt 0]], socket 1[core 15[hwt 0]], socket 1[core 16[hwt 0]], socket 1[core 17[hwt 0]]: [./././././././././././.][B/B/B/B/B/B/./././././.]
[n-62-21-38:19497] MCW rank 3 bound to socket 1[core 18[hwt 0]], socket 1[core 19[hwt 0]], socket 1[core 20[hwt 0]], socket 1[core 21[hwt 0]], socket 1[core 22[hwt 0]], socket 1[core 23[hwt 0]]: [./././././././././././.][././././././B/B/B/B/B/B]
```
which is the desired outcome :)

This can of course be controlled in LSF via 
```bash
#BSUB -n 4
#BSUB -R "affinity[core(6,same=socket)]"
mpirun ./run.sh
```
but that is not what I wanted. :)

Bottom line (my guess):
- OpenMPI reads the `LSB_AFFINITY_HOSTFILE` and assigns MPI processors according to `-n` for each line in the file.
- OpenMPI does not obey mappings on the command-line if `LSB_AFFINITY_HOSTFILE` is defined.

**Comment by jjhursey** [comment]
I raised this with the LSF team internally, but I think your analysis is correct. Open MPI should ignore the affinity file if the user expresses any bind/place/order options on the command line.

**Comment by zerothi** [comment]
Ok, thanks! :)

**Comment by zerothi** [comment]
Note that even when doing:
```bash
mpirun <options> -x LSB_AFFINITY_HOSTFILE= ./exec
# or
mpirun <options> -x LSB_AFFINITY_HOSTFILE="" ./exec
```
fails. I.e. even a null-string will not work... :(

Also, any progress on this?

**Comment by zerothi** [comment]
@jjhursey I don't know how I could unassign a person! Could you please add him again? Thanks! :)

**Comment by zerothi** [comment]
Yet another discovery. Env-vars are not exported before parsing:
```bash
old_LSB_AFFINITY_HOSTFILE=$LSB_AFFINITY_HOSTFILE
unset LSB_AFFINITY_HOSTFILE
mpirun -x LSB_AFFINITY_HOSTFILE=$old_LSB_AFFINITY_HOSTFILE
```
does not work.

@jjhursey should this be submitted as a new bug?

**Comment by rhc54** [comment]
> Yet another discovery. Env-vars are not exported before parsing:
> 
> ```shell
> old_LSB_AFFINITY_HOSTFILE=$LSB_AFFINITY_HOSTFILE
> unset LSB_AFFINITY_HOSTFILE
> mpirun -x LSB_AFFINITY_HOSTFILE=$old_LSB_AFFINITY_HOSTFILE
> ```
> 
> does not work.

I'm not sure I understand what you are saying here. The `-x` option states that we will set that envar to the specified value on the backend. It was expressly defined that way to allow users to set the value to something different for their applications. If you want `mpirun` to see that value, then you need to leave it in the environment.

**Comment by rhc54** [comment]
@zerothi Please give #6869 a try - it should solve the problem. I will then backport it to the release branches.

**Comment by zerothi** [comment]
@rhc54 thanks for trying out (I am running tests now).

And sorry for the env-var bug. It behaves as documented. Sorry.
